### ISSUE TITLE
Feature request: support heterogeneous wind direction inputs

### ISSUE THREAD

**Issue by pjireland** [NREL/floris#441]
**Is your feature request related to a problem? Please describe.**
I would like to be able to use FLORIS to simulate the waking behavior where both ambient wind speeds and ambient wind directions have spatial variations over the flow domain. Is this currently supported in FLORIS 3.1?

**Describe the solution you'd like**
With FLORIS 2.4, I used an interface like this:
```
FlorisInterface.reinitialize_flow_field(
    wind_speeds=..., # wind speed measurements for the points in wind_layout
    wind_directions=..., # wind direction measurements for the points in wind_layout
    ...
    wind_layout=..., # specify the locations where the wind measurements are taken
    layout_array=..., # specify the locations where the turbines are located in the flow
)
```
I'm not seeing a similar interface in FLORIS 3.1.

**Describe alternatives you've considered**
1. I looked into the `reinitialize` call, but this only supports the `layout` parameter, which seems to be the coordinate locations of the turbines. In addition, from what I understand, each wind speed and wind direction is assumed to apply to the entire flow domain, and it will run for all possible combinations of these.

2. I saw that you can specify a `het_map` parameter when instantiating a `FlorisInterface` object. However, this seems to apply only to wind speeds and not wind directions. Furthermore, I don't see a way to modify this after an object has been instantiated.

**Comment by bayc** [comment]
You are correct, we don't currently have a similar infeterface in FLORIS 3.1 to what we had in FLORIS 2.4. We do support heterogeneous wind speeds, as you noted, but not currently wind directions. With the vectorization and speed advancements of the code, the heterogeneous wind direction functionality adds additional complexity beyond the heterogeneous wind speeds. Additionally, the development team was aware of users using the het. wind speeds, but hadn't heard of folks using the het. wind directions. For these reasons, we decided to prioritze wind speeds first. Het. wind directions could possibly be added, but we don't currently have a clear road map for that. Is that something you would be interested in helping with?

To confirm/clarify your alternatives:

1. `reinitialize` does support the layout parameter, as well as updates to the wind directions and wind speeds. Currently, the wind directions and wind speeds are expanded with one another to represent a wind rose type of calculation. There is a feature implemented in the develop branch currently that allows the user to set a `time_series` flag in the `reinitialize` call that requires the wind direction and wind speed arrays to be of the same length, in which the code will then compute each wind condition specified without expanded the two arrays. This will be merged into the main branch soon in a v3.2 update.
2. You are correct, the `het_map` is currently just setup for wind speed ups and not directional turnings. And yes, it is meant to be set when the object is initialized and doesn't currently have a way to be modified. This is partly due to the redesign implemented in v3 in which we are moving towards a more explicit execution model and having less carryover of parameters/variables that can result in difficult debugging situations.

**Comment by pjireland** [comment]
Thanks for the response. To follow up on your questions/comments:
 
> Het. wind directions could possibly be added, but we don't currently have a clear road map for that. Is that something you would be interested in helping with?

My expectation is that being able to specify heterogeneous wind direction inputs will be useful for what I'm doing, but I'll do some investigation to confirm it is useful and worth adding the additional functionality. If so, I may be able to help with the implementation.

> You are correct, the het_map is currently just setup for wind speed ups and not directional turnings. And yes, it is meant to be set when the object is initialized and doesn't currently have a way to be modified.

Understood. Would it work if `het_map` is allowed to be an input to `reinitialize` (similar to `wind_speeds`, `wind_directions`, `layout`, `turbine_type`, etc.)? If `None`, the behavior could be the same as the current usage, and if specified, `het_map` could be modified to be the new specified value. I'm happy to submit a PR for this if so.

**Comment by bayc** [comment]
Yea, I think that could work. It is currently re-assigned in `reinitialize` any ways. One of the intricacies for `het_map` is that you need to define speed ups for each wind direction, so if you change the number of wind directions, you need to also generate a new `het_map` interpolant that matches that dimension, so that is one catch that we would want to check for potentially.

**Comment by pjireland** [comment]
> One of the intricacies for `het_map` is that you need to define speed ups for each wind direction, so if you change the number of wind directions, you need to also generate a new `het_map` interpolant that matches that dimension, so that is one catch that we would want to check for potentially.

Makes sense. I've created https://github.com/NREL/floris/pull/453 to allow `het_map` to be modified when `reinitialize` is called, but haven't tested for the potential issue you raised here.

**Comment by pjireland** [comment]
To follow up on some points raised previously:

> My expectation is that being able to specify heterogeneous wind direction inputs will be useful for what I'm doing, but I'll do some investigation to confirm it is useful and worth adding the additional functionality.

From my investigation, I've concluded that the ability to specify heterogeneous wind direction inputs is needed for accurate wake steering simulations of flows that we are observing in the field, as we observe significantly different optimum yaw angles and wake steering energy gain predictions depending on the assumptions made about the background wind direction field. (Incidentally, it seemed much less critical for us to capture heteogeneity in the wind speed field.)

> If so, I may be able to help with the implementation.

I'm not sure at the moment if this is something I'll be able to help implement. Is there currently a timeline for when this feature is expected to be integrated into FLORIS version 3?

**Comment by paulf81** [comment]
I think all the issues above are addressed in the final version of pull request #453 so closing this issue
